enhancements/insights/pulling-data-from-ocm.md

```diff
@@ -0,0 +1,133 @@
+---
+title: pulling-and-exposing-data-from-ocm
+authors:
+  - "@tremes"
+reviewers:
+  - "@sbose78"
+  - "@inecas"
+  - "@petli-openshift"
+  - "@smarterclayton"
+approvers:
+  - "@sbose78"
+  - "@smarterclayton"
+creation-date: 2021-03-04
+last-updated: 2021-03-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Insights Operator pulling and exposing data from the OCM API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement will enable the Insights Operator to pull the data (SCA certs)
+from the OCM (OpenShift Cluster Manager) API. The data will be exposed by the Insights Operator
+in the OpenShift API to allow users to use them when consuming and building container images
+on the platform.
+
+## Motivation
+
+Users could consume RHEL content and container images using the RHEL subscription in the OpenShift 3.x.
+In the OpenShift 4, this is no longer possible because the Red Hat Enterprise Linux Core OS (RHCOS) does not
+provide any attached subscription. This enhancement is to provide users the Simple Content Access (SCA) certs
+from Red Hat Subscription Manager (RHSM).
+
+### Goals
+
+- Extend the Insights Operator config with an OCM API URL to be able to query the data
+- Periodically pull the data from the OCM API and expose it in the OpenShift API
+
+### Non-Goals
+
+- Insights Operator providing any transformation or post-processing of the data pulled
+  from the OCM API
+
+## Proposal
+
+### User Stories
+
+#### Consume SCA certs exposed in the API
+
+As an OpenShift user
+I want to consume SCA certs to be able to consume RHEL content and to build
+corresponding container images.
+
+### Risks and Mitigations
+
+#### OCM API is down
+
+Risk: OCM API is down or doesn't provide up to date data.
+
+Risk: Insights Operator is unable to expose/update the data in the OpenShift API
+
+Mitigation: Introduce a new state in the Insights Operator (e.g "SCADataDegraded") and
```

Comment:
does this mean the clusters won't be able to upgrade if the OCM is down?

Reply:
No. This would be a new clusteroperator condition and `SCACertsExpired` is probably slightly better name.

Reply:
Keep in mind SRE toil. The operator should use an exponential backoff to retry updating the SCA cert from OCM, and an alert should only fire after a sufficient number of retries.

For comparison - the Samples Operator has an alerting mechanism if imagestreams fail to import. The alert only fires if there are failing imagestream imports after 2 hours, and the status doesn't result in the cluster being marked Degraded.

Reply:
Yes good points. There's no plan to mark a cluster as degraded. My idea was to introduce a new clusteroperator condition and create a new alert (with some threshold for number of retries) based on this condition. I think the alert will be a next step.

Reply:
i find it somewhat weird we would not mark it as degraded.  the definition of degraded is that the operator isn't providing all the function it's supposed to provide, but isn't totally broken.  Adding a new condition that amounts to saying that same thing, rather than using the condition that exists for the purpose seems odd.

It's always possible to upgrade a "Degraded" cluster(via force/override, though I recently learned that OSD is very reluctant to let you do that), but you want to do it with awareness that you're choosing to do so (imagine all your pods that get restarted after upgrade need SCACerts for some init step....wouldn't you want to know before you upgrade that that's going to be a problem?)

i think this is something we can revisit as the feature moves from techpreview to GA, but i do think we should discuss it a bit more.

Reply:
OK I didn't know that. My main reason was I didn't want to block an upgrade because of failing SCACerts pull. I think I am fine with changing to "Degraded" condition then. Thanks

Reply:
The degraded state is mentioned in the EP and will be implemented within https://issues.redhat.com/browse/CCXDEV-5227. Resolving the conversation.